drivers/ethernet/eth_ivshmem.c

```diff
@@ -0,0 +1,594 @@
+			/* Peer is not ready for init */
+			break;
+		}
+		eth_ivshmem_queue_reset(&dev_data->ivshmem_queue);
+		eth_ivshmem_set_state(dev_data, ETH_IVSHMEM_STATE_READY);
+		break;
+	case ETH_IVSHMEM_STATE_READY:
+	case ETH_IVSHMEM_STATE_RUN:
+		switch (peer_state) {
+		case ETH_IVSHMEM_STATE_RESET:
+			net_eth_carrier_off(dev_data->iface);
+			eth_ivshmem_set_state(dev_data, ETH_IVSHMEM_STATE_RESET);
+			break;
+		case ETH_IVSHMEM_STATE_READY:
+		case ETH_IVSHMEM_STATE_RUN:
+			if (dev_data->enabled && dev_data->state == ETH_IVSHMEM_STATE_READY) {
+				eth_ivshmem_set_state(dev_data, ETH_IVSHMEM_STATE_RUN);
+				net_eth_carrier_on(dev_data->iface);
+			} else if (!dev_data->enabled && dev_data->state == ETH_IVSHMEM_STATE_RUN) {
+				net_eth_carrier_off(dev_data->iface);
+				eth_ivshmem_set_state(dev_data, ETH_IVSHMEM_STATE_RESET);
+			}
+			break;
+		}
+		break;
+	}
+}
+
+FUNC_NORETURN static void eth_ivshmem_thread(void *arg1, void *arg2, void *arg3)
+{
+	const struct device *dev = arg1;
+	struct eth_ivshmem_dev_data *dev_data = dev->data;
+
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	while (true) {
+		k_sem_take(&dev_data->int_sem, K_FOREVER);
+
+		eth_ivshmem_state_update(dev_data);
+		if (dev_data->state != ETH_IVSHMEM_STATE_RUN) {
+			continue;
+		}
+
+		while (true) {
+			struct net_pkt *pkt = eth_ivshmem_rx(dev_data);
+
+			if (pkt == NULL) {
+				break;
+			}
+
+			if (net_recv_data(dev_data->iface, pkt) < 0) {
+				/* Upper layers are not ready to receive packets */
+				net_pkt_unref(pkt);
+			}
+
+			k_yield();
+		};
+	}
+}
+
+static void eth_ivshmem_isr(const void *arg)
+{
+	const struct device *dev = arg;
+	struct eth_ivshmem_dev_data *dev_data = dev->data;
+
+	k_sem_give(&dev_data->int_sem);
+}
+
+static uint64_t pcie_conf_read_u64(pcie_bdf_t bdf, unsigned int reg)
+{
+	uint64_t lo = pcie_conf_read(bdf, reg);
+	uint64_t hi = pcie_conf_read(bdf, reg + sizeof(uint32_t)/4);
+
+	return hi << 32 | lo;
+}
+
+static bool eth_ivshmem_pci_lookup_cb(pcie_bdf_t bdf, pcie_id_t id, void *cb_data)
+{
+	struct pcie_dev *pci_dev = cb_data;
+
+	if (id != PCI_ID_SIEMENS_IVSHMEM) {
+		return true;
+	}
+
+	uint32_t class = pcie_conf_read(bdf, PCIE_CONF_CLASSREV);
+
+	if (PCIE_CONF_CLASSREV_PROGIF(class) != IVSHMEM_PROTO_NET) {
+		return true;
+	}
+
+	pci_dev->bdf = bdf;
+
+	return false;
+}
+
+int eth_ivshmem_initialize(const struct device *dev)
+{
+	struct eth_ivshmem_dev_data *dev_data = dev->data;
+	const struct eth_ivshmem_cfg_data *cfg_data = dev->config;
+	int res;
+
+	struct pcie_dev pcie = {
+		.bdf = PCIE_BDF_NONE,
+		.id = PCI_ID_SIEMENS_IVSHMEM
+	};
+	struct pcie_scan_opt pcie_scan_opt = {
+		.cb = eth_ivshmem_pci_lookup_cb,
+		.cb_data = &pcie,
+		.flags = (PCIE_SCAN_RECURSIVE | PCIE_SCAN_CB_ALL),
+	};
+
+	pcie_scan(&pcie_scan_opt);
+
+	if (pcie.bdf == PCIE_BDF_NONE) {
+		LOG_ERR("Failed to find PCIe device");
+		return -ENODEV;
+	}
+
+	LOG_INF("PCIe: ID 0x%08X, BDF 0x%X", pcie.id, pcie.bdf);
```

Comment:
all the above should not be there. See DEVICE_PCIE_INST_DECLARE() instead

Reply:
Currently that functionality only allows devices to be found using a vendor/device-id tuple.

There can be multiple ivshmem PCI devices present in the controller, only those with `IVSHMEM_PROTO_NET` in the programming interface register (ProgIf) are suitable for this driver.

I used a PCI scan to do this, the alternative would be to add additional device lookup parameters to the PCI controller

Reply:
@tbursztyka I just added PCI device filtering on the class-rev register to tidy this up